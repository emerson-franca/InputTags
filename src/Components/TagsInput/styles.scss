.MuiInputBase-root {
  display: flex;
  align-items: flex-start;
  flex-wrap: wrap;
  padding: 20px;
  border-radius: 6px;
  overflow: hidden;

  background: #fff !important;
}

.MuiInputBase-input {
  padding: 10px;
}

.tags {
  border-radius: inherit !important;
  border-right: 6px solid #a5c7ea !important;
}

// decidi utilizar important ao invés do makeStyles
// para diminuir a "poluição" no código
