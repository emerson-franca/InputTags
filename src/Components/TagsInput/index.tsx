import React, { useState } from 'react'
import Chip from '@material-ui/core/Chip'
import TextField, { OutlinedTextFieldProps } from '@material-ui/core/TextField'
import Snackbar from '@material-ui/core/Snackbar'
import SnackbarContent from '@material-ui/core/SnackbarContent'

import './styles.scss'
import useEmail from '@/hooks/useEmail'
import { Button } from '@material-ui/core'

export default function TagsInput({ ...other }: OutlinedTextFieldProps) {
  const [inputValue, setInputValue] = useState('')
  const [showAlert, setShowAlert] = useState(false)
  const [emails, setEmails, sendEmail] = useEmail()

  const isValidEmail = /^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/

  const handleDelete = (item: string) => {
    const newItems = emails.filter((tag) => item !== tag)
    setEmails(newItems)
  }

  const handleTabOrEnter = (event: React.KeyboardEvent<HTMLDivElement>, inputValue: string) => {
    event.preventDefault()
    const allValues = inputValue.split(';')
    const emailsToInclude = allValues.filter(
      (value) => isValidEmail.test(value) && emails.indexOf(value) === -1
    )
    if (emailsToInclude.length > 0) {
      setEmails([...emails, ...emailsToInclude])
    } else {
      setShowAlert(true)
    }
    setInputValue('')
  }

  const handleBackspace = (inputValue: string) => {
    if (inputValue === '') {
      const ItemsAfterDelete = emails.filter((tag) => tag !== emails[emails.length - 1])
      setEmails(ItemsAfterDelete)
    }
  }

  const handleInputValue = (event) => {
    setInputValue(event.target.value)
  }

  return (
    <>
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
        open={showAlert}
        autoHideDuration={2000}
        onClose={() => setShowAlert(false)}
      >
        <SnackbarContent message="Email invÃ¡lido" />
      </Snackbar>
      <TextField
        value={inputValue}
        onChange={(event) => handleInputValue(event)}
        classes={{ root: 'text_field' }}
        onKeyDownCapture={(event) => {
          let inputValue = (event.target as HTMLTextAreaElement).value
          if (event.key === 'Tab' || event.key === 'Enter') {
            handleTabOrEnter(event, inputValue)
          }

          if (event.key === 'Backspace') {
            handleBackspace(inputValue)
          }
        }}
        InputProps={{
          'data-testid': 'tags-input',
          fullWidth: true,
          startAdornment: emails?.map((item) => (
            <Chip
              classes={{ root: 'tags' }}
              color="primary"
              key={item}
              tabIndex={-1}
              label={item}
              onDelete={() => handleDelete(item)}
              style={{ margin: '3px' }}
            />
          )),
        }}
        {...other}
      />
      <Button
        style={{
          height: '50px',
          margin: '20px 0',
        }}
        title="submit"
        variant="contained"
        color="primary"
        value="submit"
        onClick={() => sendEmail()}
      >
        Enviar
      </Button>
    </>
  )
}
