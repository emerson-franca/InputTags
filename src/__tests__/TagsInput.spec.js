import React from 'react'
import { cleanup, render, screen } from '@testing-library/react'
import { act } from '@testing-library/react-hooks'
import TagsInput from '../Components/TagsInput'
import userEvent from '@testing-library/user-event'

describe('TagsInput Component', () => {
  afterEach(jest.clearAllMocks)
  afterEach(cleanup)

  beforeAll(() => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve({ emails: ['teste@teste.com', 'teste@teste2.com'] }),
      })
    )
  })

  it('cria o snapshot do component', async () => {
    await act(async () => {
      const container = render(<TagsInput />)
      expect(container.asFragment()).toMatchSnapshot()
    })
  })

  it('deve renderizar varias tags separadas por ;', async () => {
    const container = render(<TagsInput />)

    const InputTag = screen.getByTestId('tags-input')
    userEvent.type(InputTag, 'teste@enter.com;teste@enter2.com;teste@enter3.com')
    userEvent.type(InputTag, '{enter}')

    expect(screen.getByText('teste@enter.com')).toBeInTheDocument()
    expect(screen.getByText('teste@enter2.com')).toBeInTheDocument()
    expect(screen.getByText('teste@enter3.com')).toBeInTheDocument()
  })

  it('deve renderizar tags quando preencher o input e pressionar enter', async () => {
    const container = render(<TagsInput />)

    const InputTag = screen.getByTestId('tags-input')
    userEvent.type(InputTag, 'teste@enter.com')
    userEvent.type(InputTag, '{enter}')

    expect(screen.getByText('teste@enter.com')).toBeInTheDocument()
  })

  it('deve renderizar tags quando preencher o input e pressionar tab', () => {
    const container = render(<TagsInput />)

    const InputTag = screen.getByTestId('tags-input')
    userEvent.type(InputTag, 'teste@tab.com')
    userEvent.tab()

    expect(screen.getByText('teste@tab.com')).toBeInTheDocument()
  })

  it('deve deletar a útima tag criada ao pressionar o botão de backspace', async () => {
    const container = render(<TagsInput />)

    const InputTag = screen.getByTestId('tags-input')
    userEvent.type(InputTag, 'teste@enter2.com')
    userEvent.type(InputTag, '{enter}')

    const email = screen.queryByText('teste@tab2.com')

    expect(container.asFragment()).toMatchSnapshot()
    expect(email).not.toBeInTheDocument()
  })
})
