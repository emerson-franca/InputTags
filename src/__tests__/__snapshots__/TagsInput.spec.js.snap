// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsInput Component cria o snapshot do component 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root text_field"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
      data-testid="tags-input"
    >
      <div
        class="MuiChip-root tags MuiChip-colorPrimary MuiChip-deletableColorPrimary MuiChip-deletable"
        role="button"
        style="margin: 3px;"
        tabindex="-1"
      >
        <span
          class="MuiChip-label"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon MuiChip-deleteIconColorPrimary"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
      </div>
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    style="height: 50px; margin: 20px 0px;"
    tabindex="0"
    title="submit"
    type="button"
    value="submit"
  >
    <span
      class="MuiButton-label"
    >
      Enviar
    </span>
  </button>
</DocumentFragment>
`;

exports[`TagsInput Component deve deletar a útima tag criada ao pressionar o botão de backspace 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root text_field"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
      data-testid="tags-input"
    >
      <div
        class="MuiChip-root tags MuiChip-colorPrimary MuiChip-deletableColorPrimary MuiChip-deletable"
        role="button"
        style="margin: 3px;"
        tabindex="-1"
      >
        <span
          class="MuiChip-label"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon MuiChip-deleteIconColorPrimary"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
      </div>
      <div
        class="MuiChip-root tags MuiChip-colorPrimary MuiChip-deletableColorPrimary MuiChip-deletable"
        role="button"
        style="margin: 3px;"
        tabindex="-1"
      >
        <span
          class="MuiChip-label"
        >
          teste@enter2.com
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon MuiChip-deleteIconColorPrimary"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
      </div>
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    style="height: 50px; margin: 20px 0px;"
    tabindex="0"
    title="submit"
    type="button"
    value="submit"
  >
    <span
      class="MuiButton-label"
    >
      Enviar
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
